pandas/sparse/array.py

```diff
@@ -322,31 +332,51 @@ def take(self, indices, axis=0):
             # return scalar
             return self[indices]
 
-        indices = np.atleast_1d(np.asarray(indices, dtype=int))
-
-        # allow -1 to indicate missing values
+        indices = com._ensure_platform_int(indices)
         n = len(self)
-        if ((indices >= n) | (indices < -1)).any():
-            raise IndexError('out of bounds access')
+        if allow_fill and fill_value is not None:
+            # allow -1 to indicate self.fill_value,
+            # self.fill_value may not be NaN
+            if (indices < -1).any():
+                msg = ('When allow_fill=True and fill_value is not None, '
+                       'all indices must be >= -1')
+                raise ValueError(msg)
+            elif (n <= indices).any():
+                msg = 'index is out of bounds for size {0}'
+                raise IndexError(msg.format(n))
+        else:
+            if ((indices < -n) | (n <= indices)).any():
+                msg = 'index is out of bounds for size {0}'
+                raise IndexError(msg.format(n))
+
+        is_empty = True
 
         if self.sp_index.npoints > 0:
-            locs = np.array([self.sp_index.lookup(loc) if loc > -1 else -1
-                             for loc in indices])
-            result = self.sp_values.take(locs)
-            mask = locs == -1
+            if allow_fill and fill_value is not None:
+                # leave -1 as missing
+                locs = np.array([loc if loc != -1 else loc
+                                 for loc in indices])
+            else:
+                # convert negative indices to positive
+                locs = np.array([loc if loc >= 0 else n + loc
+                                 for loc in indices])
+            locs = [self.sp_index.lookup(loc) for loc in locs]
+            locs = np.array(locs)
```

Comment:
should (maybe make an issue), to have a ``.lookup_array(int[:])`` for this, but low-priortiy